### PR TITLE
as_strided batching rule

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #47227 Batched gradient support for view+inplace operations
* #47365 Add batching rule for torch.clone(tensor, torch.contiguous_format)
* **#47364 as_strided batching rule**
* #47226 Batching rule for Tensor.new_empty_strided
* #47225 Implement Tensor.new_empty_strided(sizes, strides, *, dtype, device, requires_grad)

This PR adds a batching rule for as_strided. `as_strided` is a really weird
operation and I hope that users don't use it very much.

Motivation
----------
The motivation for adding a batching rule for as_strided is for
batched gradient computation.

AsStridedBackward appears in PyTorch when handling view+in-place
operations and calls `as_strided`. AsStridedBackward calls as_strided on
a fresh tensor with storage_offset equal to 0. We would like to be able
to vmap through the backward graph of view+in-place operations to
for batched gradient computation, especially because internally we have
a number of functions that are implemented as a view+in-place.

Alternatives
------------
If we think that as_strided is too crazy to have a batching rule, we
could either:
- have a flag that controls the autograd view+in-place
behavior
- require that the input tensor's storage offset must be equal to 0
to make it easier to reason about.

I think the batching rule makes sense, so I didn't pursue the
alternatives.

The batching rule
-----------------
```
y = vmap(lambda x: x.as_strided(sizes, strides, offset))(xs)
```
The result of the above should be "equivalent" to:
- Assume that each x has storage offset equal to xs.storage_offset()
(call that S).
- Calling as_strided with (sizes, sizes, offset + x[i].storage_offset() - S) on each x.

More concretely,
this returns a view on `xs`, such that each y[i] has:
- sizes: `sizes`
- strides: `strides`
- storage_offset: offset + i * x.stride(batch_dim)

Why the behavior can be weird
-----------------------------
The behavior of the batching rule may be different from actually running
as_strided in a for-loop because `as_strided` takes in `offset` as a
"absolute offset". As an example, consider

```
>>> x = torch.tensor([0., 1., 2., 3., 4.])
>>> z = [x[i].as_strided([1], [1], 0) for i in range(5)]
```
Each z[i] is actually the same view on x (z[i] == torch.tensor([0.]))!
However, we consider the above for-loop comprehension to be a user error:
a user should have written the following if they wanted to use as_strided
in a per-sample way:
```
>>> z = [x[i].as_strided([1], [1], 0 + x[i].storage_offset()) for i in range(5)]
```

Test Plan
---------
- Added some tests that compare vmap+as_strided to vmap+(the equivalent operator)

Differential Revision: [D24741685](https://our.internmc.facebook.com/intern/diff/D24741685)

### CODE REVIEW

**Comment by zou3519** [comment]
What changed from #47224:
- [Corrected strides() call](https://github.com/pytorch/pytorch/pull/47224#discussion_r516703533)
- Explained the ["without loss of generality"](https://github.com/pytorch/pytorch/pull/47224#discussion_r516720150)
- Added some [sanity checks](https://github.com/pytorch/pytorch/pull/47224#discussion_r516720150)
- Fixed some nits
- Made a [note](https://github.com/pytorch/pytorch/pull/47224#pullrequestreview-522544748) to add a note about as_strided to a vmap gotchas doc (I have this local text file where I am storing the gotchas things I come across)

File: aten/src/ATen/BatchingRegistrations.cpp [hunk 1]
```diff
@@ -354,6 +354,234 @@ Tensor view_as_complex_batching_rule(const Tensor& self) {
...
+  auto slice_sizes = physical_tensor.sizes().slice(num_batch_dims);
+  auto slice_strides = physical_tensor.strides().slice(num_batch_dims);
+  auto base_offset = physical_tensor.storage_offset();
+
+  auto storage_offset = maybe_storage_offset.value_or(base_offset);
+
+  auto max_as_strided_loc = maximum_indexable_location(sizes, strides, storage_offset);
+  auto max_slice_loc = maximum_indexable_location(slice_sizes, slice_strides, base_offset);
+
+  if (!max_as_strided_loc.has_value() && !max_slice_loc.has_value()) {
```
Comment (albanD) [inline aten/src/ATen/BatchingRegistrations.cpp:408]: This if and the one below is equivalent to `if (!max_as_strided_loc.has_value()) {` no?

Comment (zou3519) [reply aten/src/ATen/BatchingRegistrations.cpp]: Fixed!

File: aten/src/ATen/BatchingRegistrations.cpp [hunk 2]
```diff
@@ -354,6 +354,231 @@ Tensor view_as_complex_batching_rule(const Tensor& self) {
...
+
+  auto storage_offset = maybe_storage_offset.value_or(base_offset);
+
+  auto max_as_strided_loc = maximum_indexable_location(sizes, strides, storage_offset);
+  auto max_slice_loc = maximum_indexable_location(slice_sizes, slice_strides, base_offset);
+
+  if (!max_as_strided_loc.has_value()) {
+    return;
+  }
+  if (max_as_strided_loc.has_value() && !max_slice_loc.has_value()) {
```
Comment (albanD) [inline aten/src/ATen/BatchingRegistrations.cpp:411]: nit: the check for `max_as_strided_loc` having value is not needed here because of the one above.

**Review by albanD — APPROVED** [review]
Thanks for all the comments! They definitely help following what's happening here!
